pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,917 @@
+		}
+
+		instanceInfo, err = recon.reconstructInstanceInfo(iiName, instance, instanceInfo)
+		if err != nil {
+			klog.Errorf("Error creating InstanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+
+		instanceInfo, err = recon.maybeUpdateInstanceInfoStatus(instance, instanceInfo, instanceShares)
+		if err != nil {
+			klog.Errorf("Error updating instanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+	}
+
+	return instanceInfoMap
+}
+
+// createAndUpdateShareInfos create shareInfo objects if needed and updates their statuses to match with reality.
+// ShareInfo objects in the returned map must be treated as read only.
+func (recon *multishareReconciler) createAndUpdateShareInfos(shares []*file.Share) map[string]*v1alpha1.ShareInfo {
+	shareInfoMap := make(map[string]*v1alpha1.ShareInfo)
+
+	// Create ShareInfo that are not reflected.
+	for _, share := range shares {
+		shareInfoName := util.ShareToShareInfoName(share.Name)
+		shareInfo, err := recon.shareLister.Get(shareInfoName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("Error getting shareInfo %q from informer: %v", shareInfoName, err)
+				continue
+			}
+			shareInfo, err = recon.clientset.MultishareV1alpha1().ShareInfos().Get(context.TODO(), shareInfoName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					klog.Errorf("Error getting shareInfo %q from api server: %v", shareInfoName, err)
+					continue
+				}
+				// shareInfo does not exist for share
+				klog.V(4).Infof("ShareInfo object for share %q not found in API server", shareInfoName)
+				shareInfo = nil
+			}
+			// shareInfo exist in api server but not cache
+			klog.V(4).Infof("ShareInfo object for share %q not found in informer cache but found in api server", shareInfoName)
+		}
+
+		shareInfo, err = recon.createShareInfo(share, shareInfo)
+		if err != nil {
+			klog.Errorf("Error creating ShareInfo %q: %v", shareInfoName, err)
+			continue
+		}
+		shareInfoMap[shareInfoName] = shareInfo
+
+		shareInfo, err = recon.maybeUpdateShareInfoStatus(share, shareInfo)
+		if err != nil {
+			klog.Errorf("Error updating ShareInfo %q: %v", shareInfoName, err)
+			continue
+		}
+		shareInfoMap[shareInfoName] = shareInfo
+	}
+
+	return shareInfoMap
+}
+
+func (recon *multishareReconciler) reconstructInstanceInfo(iiName string, instance *file.MultishareInstance, instanceInfo *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, error) {
+	if instanceInfo != nil {
+		return instanceInfo, nil
+	}
+
+	instanceInfo = &v1alpha1.InstanceInfo{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: iiName,
+			Labels: map[string]string{
+				driver.ParamMultishareInstanceScLabel: instance.Labels[util.ParamMultishareInstanceScLabelKey],
+			},
+		},
+		Spec: v1alpha1.InstanceInfoSpec{
+			CapacityBytes: instance.CapacityBytes,
```

Comment:
In generateInstanceInfo() we put the storage class name in spec.StorageClassName, don't we need it here?

Reply:
we don't, because the only need for `StorageClassName` is to get parameter when calling Filestore API to create a new instance. since Instance already exist here, there's no need for StorageClassName here.

Reply:
Please address the reason as a comment in the followup PR.

Reply:
in the reconstruct case, the instance is already present with a `scTag`. And we cannot change the instances' property anymore. we don't need to have `storageClassName` field (the storageclass might not exist either) and it will act like a distinguisher between migrated instances and driver-created instances.